### PR TITLE
[DNM] Add ARC cache support and refactor cache system

### PR DESCRIPTION
**Pull request by blacktear23**
* Add an ARC cache implements based on LRU.
* Refactor cache system introduce Cache interface for ARC and LRU.
* Use a variable to change default cache type as ARC or LRU

### CODE REVIEW

File: server/cache/cache_test.go [hunk 1]
```diff
@@ -178,3 +178,79 @@ func (s *testRegionCacheSuite) TestFifoCache(c *C) {
...
+
+	val, ok = cache.Get(1)
+	c.Assert(ok, IsTrue)
+	c.Assert(val, DeepEquals, "1")
+
+	cache.Put(4, "4")
+	cache.Put(4, "4")
+
+	c.Assert(cache.Len(), Equals, 3)
+
```
Comment (nolouch) [inline server/cache/cache_test.go]: better to add a comment like:
```
_ _ _ _ | _ _ _ _ | 4 1 2 _ | 3 _ _ _ |
    b1       t1        t2        b2
```

File: server/cache/arc.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+				c.replace(true)
+			}
+		}
+		return
+	}
+
+	// Put to the recently seen list
+	c.t1.Put(key, value)
+
+	// Potentially need to make room in the cache
```
Comment (nolouch) [inline server/cache/arc.go]: better to extract as a function

File: server/cache.go [hunk 3]
```diff
@@ -346,15 +346,15 @@ type clusterInfo struct {
...
+	writeStatistics cache.Cache
 }
 
 func newClusterInfo(id IDAllocator) *clusterInfo {
 	return &clusterInfo{
 		id:              id,
 		stores:          newStoresInfo(),
 		regions:         newRegionsInfo(),
-		writeStatistics: cache.NewLRU(writeStatLRUMaxLen),
+		writeStatistics: cache.NewDefaultCache(writeStatLRUMaxLen),
```
Comment (nolouch) [inline server/cache.go]: maybe ARC more suitable for our `writerStatistics`, but we just need to peek the frequent cache t2. cc @disksing

File: server/cache/cache.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+// NewCache create cache by type
+func NewCache(size int, cacheType int) Cache {
+	switch cacheType {
+	case LRUCache:
+		return NewLRU(size)
+	case ARCCache:
+		return NewARC(size)
+	}
+	panic("Unknown cache type")
```
Comment (nolouch) [inline server/cache/cache.go]: better to  replace this with our log package

File: server/cache/cache_test.go (hunk 1 above)
Comment (blacktear23) [reply server/cache/cache_test.go]: I found some bug on ARC put, fix will be push later

**Comment by blacktear23** [comment]
My purpose is:
1.  provide a plugable cache framework 
2. add ARC cache engine under cache framework
3. maybe some case use ARC is better than LRU, we can support it.

For ARC cache, it is more suitable for cache hot point than LRU. For LRU such as a full scan may flush full cache to new entry. If we have hot point and random access ARC will be more efficient.

**Comment by ngaut** [comment]
Yes, please add [DNM] to the title of this PR.

File: server/cache/arc.go [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"sync"
+)
+
+// ARC is a thread-safe fixed size Adaptive Replacement Cache.
```
Comment (ngaut) [inline server/cache/arc.go]: Could you add a reference of the ARC design ?

File: server/cache/arc.go [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+// ARC is an enhancement over the standard LRU cache in that tracks both
```
Comment (ngaut) [inline server/cache/arc.go]: What's the difference of your implementation and https://github.com/moovweb/go-cache ?

**Comment by ngaut** [comment]
Could you split the PR into two parts? One for providing a plugable cache framework, and the other for the implementation of ARC.

**Comment by blacktear23** [comment]
@ngaut Sure, I will split it into two. In addition ARC cache is referenced by https://github.com/hashicorp/golang-lru/blob/master/arc.go

**Comment by disksing** [comment]
Thanks @blacktear23 , but I guess we will not accept the ARC cache mainly due to legal concerns. You may want to refer https://en.wikipedia.org/wiki/Adaptive_replacement_cache

> In 2006, IBM was granted a patent for the adaptive replacement cache policy.
> ...
> PostgreSQL used ARC in its buffer manager for a brief time (version 8.0.0), but quickly replaced it with another algorithm, citing concerns over an IBM patent on ARC.

**Comment by blacktear23** [comment]
@disksing Gotcha

**Pull request closed without merging**